### ISSUE TITLE
Prompt to open a folder after connecting to a SSH host

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-remote-release#511]
Currently when you connect to a SSH host, you land in an empty window. From here you can then open a folder or open a workspace on the remote host.

While this makes sense in principal, in talking with @chrisdias and seeing a few people do it for the first time, it's not immediately obvious what to do next. This creates a rough on-boarding experience for new users.

I'd propose we instead:
1. Connect as we do now
2. Once the connection is complete, trigger "Open Folder" on the remote host
3. If developers do not want to open a folder, they can then hit "ESC" and do something else instead.

Once you've initially opened a folder or workspace, you can access it again via the SSH explorer, so optimizing for the first time experience seems to make sense here.

//cc: @chrisdias @roblourens @kieferrm

**Comment by roblourens** [comment]
We talked about this initially but currently we don't have a way to pass some parameters through to the new window. It could be encoded in the remote authority but I don't think that's the right place for it.

**Comment by aeschli** [comment]
> If developers do not want to open a folder, they can then hit "ESC" and do something else instead.

The ESC is a problem, it will confuse. I guess we could have a dialog that lets you select files, folders or workspaces.

**Comment by Chuxel** [comment]
@roblourens, Given the effort, @aeschli brought up a good point [here](https://github.com/microsoft/vscode-remote-release/issues/501#issuecomment-497122688) that we could consider as a starting point to help orient people.

>The empty window first also solves some issues for us. One issue is that we have 3 different artifacts you can open: files/folders and workspaces, with 3 (MacOs: 2) different commands and keyboard shortcuts.
>I suggest we make the empty window experience better, e.g. by showing a welcome page with prominent open buttons.

This welcome page could also include other features people may not be aware of like port forwarding in the SSH case given we can hyperlink commands.  So it isn't annoying, we would show the welcome page the first time you get a new window for a given connection type - any SSH host or WSL, but in the SSH case we probably wouldn't need to do it for every time you connected to a new host.

We could potentially do this for containers as well, but you end up opening a folder to get the config, so it is probably less critical there. You very much open with intent in that case -- even in the remote docker case. There's also a proposed Open Workspace in Container command to address that gap (#387).

**Comment by kieferrm** [comment]
Extensions don't have a way to contribute to the welcome page today we either need to open this up or we open a custom welcome page the same way other extensions have done. If open folder is the key problem, we can for example always show the file explorer on first connect to a remote host with an empty window. There we have a big `Open Folder` button.

**Comment by Chuxel** [comment]
@kieferrm Hmmm. That might help -- but I think the main problem is it looks like your local empty window, so it's not obvious that the connection actually worked unless you know where to look.  Pointing to what should be done next would resolve that... we could fall back on a one time notification, but we know that UX is getting to be overused. A one time custom page might be more useful.

**Comment by isidorn** [comment]
+1 on @chuxel first post. I basically wrote this in another issue:

This empty VSCode window does provide any useful functionality and we want to get the user as soon as possible out of this. The only thing that the user can get out of this is that he is happily connected to remote.

I propose that once the connection is made that we automatically show a quick open dialog and offer the user to choose a remote folder to open.
I understand this is not the pattern that we commonly use, however we already have some multi step quick open actions for remote, the only difference here is that one of the steps is opening a new window.

This opening of a dialog should only be done after the user invoked the "Connect to a Host" action, not every time he is in an empty window on remote.

The alternative would be to make sure the explorer is opened, and we could tune the empty explorer text. To change the wording that it is clear the the user is opening a folder on the remote.

I simply prefer the first approach since I believe it is more straight forward to the user and we skip this state where the user can no do anything useful.

**Comment by isidorn** [comment]
To @aeschli point since we use a custom dialog we could support that the user can select anything just like on the mac.
As for the techinical difficualty of opening this remote dialog on a new window. We could support some new parameters or there might be something we could already use @bpasero

**Comment by bpasero** [comment]
There is currently no parameter to my knowledge that opens a new window and then opens a remote file dialog.

**Comment by kieferrm** [comment]
The challenge we have in Remote is the same we had in the local case and we mitigated/solved by introducing the welcome page. In the Remote case we also supports editing individual files or opening folders.  Therefore, opening a modal dialog on top of a new window to hint the user at what they could do next is not a great idea. We could
- show prominently in the welcome page that this window is connected to a remote context (but it's unclear what to do with sections like `Recent` as those are all local when you use Remote for the first time
- open the file explorer by default (with a clear indication about the remote context).

**Comment by isidorn** [comment]
@kieferrm I can look into opening the explorer by default and polishing the message there if you think that is best?

**Comment by isidorn** [comment]
@kieferrm @roblourens I tried the SSH, and have a couple of questisons / feedback
1. SSH: connect to host does not remember my previous folder, it always enters a no folder vscode. Is it possible that this action remembers the last active folder in that SSH connection? This way the user would not end up in the empty folder experience this often
2. If we decide that showing explorer is the best way is it possible for the ssh extension to simply execute the following command once the connection is established `workbench.view.explorer`
3. I could polish the empty explorer text for this case from 'You have not yet opened a folder' -> 'You are connected to an SSH remote, please open a folder'. Suggestions on wording welcome

Unrelated to that I am still not convinced that automatically opening quick open is not the best way to go here. As for Kai's argument that users can still edit individual files and open folders -> for both those activites they need the quick open first. As far as I understand the quick open supports opening both. I am simply not aware of any action a user can do that does not require first to use quick open.

Let me know what you think

**Comment by kieferrm** [comment]
@isidorn 

About 1) This is an interesting idea. We'd then miss an action to open a new empty window on a SSH box since "New Window" opens a local window. So, we can't simply change the behavior here without impacting other areas.

About 2) Sounds good. Need to find a good trigger for calling the action.

About 3) How about: "Connected to ${host} via SSH".  The "Open Folder" button would still  be right below, so I think we can omit the call to open a folder in the text.

**Comment by isidorn** [comment]
1. I think changing the behavior of that action is fine. We are changing it for the better imho and I do not see a use case we would break. The user can still exectue that action and Close Folder to get into the empty. You can also introduce a new action which would only be in the command palette and not in any prominent place. I would call it `Connect to Host with Empty Workspace...`. Also I would introduce this action only if we see users complaining
2. Cool. Let me know if you would like me to provide a PR
3. Sounds good, I can change this today since it is a good change unrelated to 1) and 2). The only questionable part of the message is `via SSH`, since that is specific for only 1 extension. I think a cleaner way would be to have something in common for all remote extensions. I'll check this in the afternoon

**Comment by kieferrm** [comment]
I made the following experimental change in the SSH extension: If the workbench is empty and no welcome page is shown then reveal the explorer and open the terminal.

**Comment by isidorn** [comment]
I have pushed the changes to the empty view in the explorer.
So let's see how this behaves. The text has no "ssh" hacks, it simply says Connected to ${host label}, the same host label which is shown in the status bar. This solution is nice since it works for all remote cases. In case the remote is not yet connected it simply says "Connecting..."

I would also suggest changing 1), though leaving that up to @kieferrm 
Let me know if I can help more.

![hello](https://user-images.githubusercontent.com/1926584/60024888-e9031700-9698-11e9-9629-b17573d82d24.gif)

**Comment by egamma** [comment]
>About 3) How about: "Connected to ${host} via SSH".

👍 but let's keep in mind that the message should also be tuned for the other authorities WSL, Containers, but also consider contributed ones (e.g. Cloud Envs).

**Comment by isidorn** [comment]
@egamma yes, my solution covers all those cases.

**Comment by Chuxel** [comment]
Yeah the early results from the interviews with full-time SSH users @bowdenk7 has been driving have indicated a 50/50 split between whether they were confused at first when they hit this or not. It was never considered a big problem, however, as they all figured it out - but each were very experienced SSH users. They also all were opening a folder when they first tried out the SSH extension (not a file or VS Code work space).  The unprompted suggestion was usually to go right to open folder as a result.  On the containers side, we need more data, but the open folder behavior it has seemed to meet with expectations.

That said, we can get a bit more insight from part-time users to understand how confusing/painful this actually is for them.  Pain was low for full-timers.  What is described here seems like a good first step to me -- we can compare the full-time feedback to part-time users as we talk with more of them formally this week and decide if we need to do more.

**Comment by gordonrust** [comment]
Is there any progress on the suggestions 1 and/or 2 of @isidorn in the comment dated June 21, 2019?
Specifically, my usecase is  to land into the previous opened folder once I ssh into a server. Is it possible in vscode? Any Suggestions?

**Comment by altimmons** [comment]
+1 on this.  Id like to be able to save a workspace.  So that it loads my config.  Doing the work I am doing requires freq. restarting of the host, which disconnects me.   

In the remote hosts tab- sometimes it shows file targets.   But not always.  I cant figure out how to add it manually.  Further its a very large number of files, I need it to read the ignore list before it loads everything, because that means I have to reload the directory a 3rd time.

Preferred behaviour might be to click a target in the Remote Explorer activity bar, and have it open in the current window..  Is there a way to make it the default behavior- opening in the current window?  I see that you can in the dropdown.
But anyway- in the remote explorer under the Hostname, should be a list of files (that can be modified!) and also include workspaces.   I dont see the difficulty of running the remote host, and then opening the workspace or folder selected.  Not opening to a NULL directory.

**Comment by ravensorb** [comment]
Agree on the save workspace idea.  It would also be nice if we could set a default folder to start with in the config file

**Comment by r0oland** [comment]
+1 from me, would be great if the remote session either restores the previous folder or opens up a default folder based on the config, etc home folder.

**Comment by direnc** [comment]
+1 for saving workspace. 
I also wonder if it is not possible to add some commands to the config file in order to run in the remote server when the connection is established. Like always opening the same folder after connection.

**Comment by Thaina** [comment]
I wish I could also specified first folder in the config

Something like 

```
Host MyLanRemoteLastProj
  HostName 192.168.1.55
  User me
  DefaultPath ./Projects/MyLastProj
```

**Comment by shijiezhou1** [comment]
> I wish I could also specified first folder in the config
> 
> Something like
> 
> ```
> Host MyLanRemoteLastProj
>   HostName 192.168.1.55
>   User me
>   DefaultPath ~/Projects/MyLastProj
> ```
> 
> And it will open that folder immediately

We would like to see this feature so far no one help on it

**Comment by Thaina** [comment]
There are suggestions to use `RemoteCommand` variable to `cd` and so on. Are there any command to focus vscode to target directory from command line though?

https://serverfault.com/a/1045158

```
Host websites-my-host
    HostName <realhostname>
    IdentityFile ~/.ssh/keyfile
    User webmaster
    RequestTTY force # check if this is necessary
    RemoteCommand cd /var/www/websites && bash -l
```

**Comment by Thaina** [comment]
It seem there is similar request for `RemoteCommand` in #6215

**Comment by VSCodeTriageBot** [comment]
This issue is being closed to keep the number of issues in our inbox on a manageable level, we are closing issues that are not going to be addressed in the foreseeable future: We look at the number of votes the issue has received and the number of duplicate issues filed. More details [here](https://aka.ms/vscode-out-of-scope). If you disagree and feel that this issue is crucial: We are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting](https://aka.ms/vscodeissuereporting) guidelines.

Thanks for your understanding and happy coding!

**Comment by roblourens** [comment]
This hasn't come up in awhile, and I think the Getting Started page with the Open Folder option helps. I think that putting up an Open Folder dialog wouldn't help.

**Comment by zs-dima** [comment]
RemoteCommand option looks even better.
Open folder dialog could confuse if you want just execute some commands in console

**Comment by masteradhoc** [comment]
+1

**Comment by yenk4** [comment]
+

**Comment by spfrommer** [comment]
+1

**Comment by youpsla** [comment]
+1

**Comment by smeisegeier** [comment]
+1

**Comment by MoreTore** [comment]
+1

**Comment by QFlachaire** [comment]
+1

**Comment by UNiXMIT** [comment]
+1

**Comment by 0xphilipp** [comment]
+1

**Comment by Ilyua** [comment]
+1

**Comment by hendra-jti** [comment]
![Image](https://github.com/user-attachments/assets/15bdeef7-8f1d-4a03-9d7a-54790eb08936)

this folder is created automatically right ? but when i access another server, those folder are not created

**Comment by jacasch** [comment]
+1

It would be super great if we could add a "DefaultDirectory" in the config like this

Host <my_name>
  IdentityFile <my_public_key>
  HostName <my_host_name>
  User <user_name>
  DefaultDirectory /home

**Comment by altimmons** [comment]
Agree with above.  Awkward implementation to land nowhere.

If you cant do the above- Default Directory key, or open folder dialog box, then at least hard code the remote client with a sensible starting place '/'  or '.../home/

**Comment by akiyamasho** [comment]
+1
